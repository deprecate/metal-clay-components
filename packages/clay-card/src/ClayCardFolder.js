import './ClayCardBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayCardFolder.soy.js';

/**
 * Metal ClayCardFolder component.
 */
class ClayCardFolder extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayCardFolder.STATE = {
	/**
	 * Actions in the dropdown on card.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {?array|undefined}
	 * @default undefined
	 */
	actionItems: Config.arrayOf(
		Config.shapeOf({
			label: Config.string().required(),
			separator: Config.bool().value(false),
			url: Config.string(),
		})
	),

	/**
	 * Transform card with checkbox.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {?array|undefined}
	 * @default undefined
	 */
	checkbox: Config.shapeOf({
		checked: Config.bool().value(false),
		disabled: Config.bool().value(false),
		indeterminate: Config.bool().value(false),
		name: Config.string(),
		value: Config.string(),
	}),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * Title of the card.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {?string|undefined}
	 * @default undefined
	 */
	title: Config.string(),

	/**
	 * Transform title in the url.
	 * @instance
	 * @memberof ClayCardFolder
	 * @type {?string}
	 * @default undefined
	 */
	url: Config.string(),
};

defineWebComponent('clay-card-folder', ClayCardFolder);

Soy.register(ClayCardFolder, templates);

export {ClayCardFolder};
export default ClayCardFolder;
