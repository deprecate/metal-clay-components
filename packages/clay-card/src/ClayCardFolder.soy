{namespace ClayCardFolder}

{template .render}
    {@param spritemap: string}
    {@param? actionItems: list<[
        label: string,
        separator: bool,
        url: string
    ]>}
    {@param? checkbox: [
		checked: bool,
		disabled: bool,
		indeterminate: bool,
		name: string,
		value: string
	]}
    {@param? elementClasses: string}
    {@param? id: string}
    {@param? title: string}
    {@param? url: string}

    {call ClayCardBase.render}
        {param actionItems: $actionItems /}
        {param checkbox: $checkbox /}
        {param elementClasses: $elementClasses /}
        {param id: $id /}
        {param ref: 'cardFolder' /}
        {param spritemap: $spritemap /}
        {param title: $title /}
        {param url: $url /}
        {param variant: 'folder' /}
    {/call}
{/template}
