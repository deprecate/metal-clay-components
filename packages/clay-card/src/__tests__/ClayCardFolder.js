import ClayCardFolder from '../ClayCardFolder';

let component;
let spritemap = 'icons.svg';
let actionItems = [
	{
		label: 'Edit',
		url: '#1',
	},
	{
		label: 'Save',
		url: '#1',
	},
];

describe('ClayCardFolder', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayCardFolder({
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with classes', () => {
		component = new ClayCardFolder({
			spritemap: spritemap,
			elementClasses: 'my-custom-class',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with id', () => {
		component = new ClayCardFolder({
			spritemap: spritemap,
			id: 'myId',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with actionItems', () => {
		component = new ClayCardFolder({
			actionItems: actionItems,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with checkbox checked', () => {
		component = new ClayCardFolder({
			checkbox: {
				checked: true,
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with checkbox disabled', () => {
		component = new ClayCardFolder({
			checkbox: {
				checked: true,
				disabled: true,
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with checkbox indeterminate', () => {
		component = new ClayCardFolder({
			checkbox: {
				indeterminate: true,
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with checkbox `name`', () => {
		component = new ClayCardFolder({
			checkbox: {
				checked: true,
				name: 'checkbox01',
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with checkbox `value`', () => {
		component = new ClayCardFolder({
			checkbox: {
				checked: true,
				value: 'checkbox',
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with title', () => {
		component = new ClayCardFolder({
			title: 'empty-background.png',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCardFolder with url', () => {
		component = new ClayCardFolder({
			url: '#1',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});
});
