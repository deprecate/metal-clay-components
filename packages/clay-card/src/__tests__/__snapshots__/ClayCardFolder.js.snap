// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayCardFolder should render a ClayCardFolder with actionsItems 1`] = `
<div class="card card-type-directory card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
      <div class="flex-col">
        <div class="dropdown dropdown-action">
          <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
              <use xlink:href="icons.svg#ellipsis-v"></use>
            </svg>
          </button>
          <div class="dropdown-menu">
            <ul class="list-unstyled">
              <li>
                <a class=" dropdown-item" href="#1">Edit</a>
              </li>
              <li>
                <a class=" dropdown-item" href="#1">Save</a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with checkbox \`name\` 1`] = `
<div class="form-check form-check-card form-check-middle-left card-type-directory">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" name="checkbox01" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with checkbox \`value\` 1`] = `
<div class="form-check form-check-card form-check-middle-left card-type-directory">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" value="checkbox" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with checkbox checked 1`] = `
<div class="form-check form-check-card form-check-middle-left card-type-directory">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with checkbox disabled 1`] = `
<div class="form-check form-check-card form-check-middle-left card-type-directory">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" disabled="disabled" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with checkbox indeterminate 1`] = `
<div class="form-check form-check-card form-check-middle-left card-type-directory">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with classes 1`] = `
<div class="card card-type-directory card-horizontal my-custom-class">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
    </div>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with id 1`] = `
<div class="card card-type-directory card-horizontal" id="myId">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
    </div>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with title 1`] = `
<div class="card card-type-directory card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters">
        <div class="card-title text-truncate" title="empty-background.png">empty-background.png</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayCardFolder should render a ClayCardFolder with url 1`] = `
<div class="card card-type-directory card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters">
        <a class=" card-title text-truncate" href="#1"></a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayCardFolder should render the default markup 1`] = `
<div class="card card-type-directory card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
    </div>
  </div>
</div>
`;
